amqp/README.md

```diff
@@ -0,0 +1,274 @@
+**Message:**
+```json
+{
+  "port": 1,                 // LoRaWAN FPort
+  "confirmed": false,        // Whether the downlink should be confirmed by the device
+  "payload_raw": "AQIDBA==", // Base64 encoded payload: [0x01, 0x02, 0x03, 0x04]
+}
+```
+
+**Usage (RabbitMQ):**
+`rabbitmqadmin publish exchange='ttn.handler' routing_key='my-app-id.devices.my-dev-id.down' payload='{"port":1,"payload_raw":"AQIDBA=="}'`
+
+**Usage (Go client):**
+```go
+package main
+
+import (
+	"github.com/TheThingsNetwork/go-utils/log/apex"
+	"github.com/TheThingsNetwork/ttn/amqp"
+	"github.com/TheThingsNetwork/ttn/core/types"
+)
+
+func main() {
+
+	ctx := apex.Stdout().WithField("test", "test")
+	c := amqp.NewClient(ctx, "guest", "guest", "localhost:5672")
+	c.Connect()
+	p := c.NewPublisher("amq.topic")
+  if err := p.Open(); err != nil {
+    ctx.WithError(err).Error("Could not open publishing channel")
+  }
+  defer p.Close()
+	d := types.DownlinkMessage{
+    AppID:      "my-app-id",
+    DevID:      "my-dev-id",
+    FPort:      1,
+    PayloadRaw: []byte{0x01, 0x02, 0x03, 0x04}}
+  p.PublishDownlink(d)
+	//...
+}
+```
+
+** Usage (Python) **
+```python
+#!/usr/bin/env python
+import pika
+import sys
+
+connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost', port=5672))
+channel = connection.channel()
+
+channel.exchange_declare(exchange='ttn.handler',
+                         type='topic')
+
+routing_key = 'my-app-id.devices.my-dev-id.down'
+message = '{"port": 1, "confirmed": false, "payload_raw": "AQIDBA=="}'
+
+channel.basic_publish(exchange='ttn.handler',
+                      routing_key=routing_key,
+                      body=message)
+print(" [x] Sent %r:%r" % (routing_key, message))
+connection.close()
+```
+
+## Device Events
+
+**Routing key:** 
+
+* `<AppID>.devices.<DevID>.events.<event>`
+* `0102030405060708.devices.abcdabcd12345678.events.activations`
+* `*.devices.*.events.*`
+
+**Message:**
+```json
+{
+  "payload": "Base64 encoded LoRaWAN packet",
+  "gateway_id": "some-gateway",
+  "config": {
+    "modulation": "LORA",
+    "data_rate": "SF7BW125",
+    "counter": 123,
+    "frequency": 868300000,
+    "power": 14
+  }
+}
+```
+
+**Usages (GO)**
+
+*Subscription:*
+```go
+package main
+
+import (
+	"github.com/TheThingsNetwork/go-utils/log/apex"
+	"github.com/TheThingsNetwork/ttn/amqp"
+	"github.com/TheThingsNetwork/ttn/core/types"
+)
+
+func main() {
+
+  ctx := apex.Stdout().WithField("test", "test")
+  c := amqp.NewClient(ctx, "guest", "guest", "localhost:5672")
+  if err := c.Connect(); err != nil {
+    ctx.WithError(err).Error("Could not connect")
+  }
+  s := c.NewSubscriber("amq.topic", "", true, false)
+  if err := s.Open(); err != nil {
+    ctx.WithError(err).Error("Could not open subcription channel")
+  }
+	err = s.SubscribeAppEvents("my-app-id", "some-event",
+			func(_ Subscriber, appID string, eventType types.EventType, payload []byte) {
+			  // Do your stuff
+			})
+	//...
+}
+```
+
+*Publish:*
```

Comment:
We won't allow applications to publish device events themselves so this example can be removed